test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,307 @@
+}
+
+// setupVMAgentTest creates ExternalNode, starts antrea-agent
+// and returns a list of VMs upon success
+func setupVMAgentTest(t *testing.T, data *TestData) ([]vmInfo, error) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.winVMs, testOptions.linVMs)
+	nsReturned, _ := data.clientset.CoreV1().Namespaces().Get(context.TODO(), nameSpace, metav1.GetOptions{})
+	saReturned, _ := data.clientset.CoreV1().ServiceAccounts(nameSpace).Get(context.TODO(), serviceAccount, metav1.GetOptions{})
+	t.Logf("Using service account %s", saReturned.Name)
+	t.Logf("Using namespace %s", nsReturned.Name)
+	var vmList []vmInfo
+	if testOptions.linVMs != "" {
+		vms := strings.Split(testOptions.linVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
+			tempVm := getVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	if testOptions.winVMs != "" {
+		vms := strings.Split(testOptions.winVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Windows VM: %s", vm)
+			tempVm := getWindowsVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	t.Logf("TestVMAgent setup")
+	for _, vm := range vmList {
+		stopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalNode for VM: %s", vm.nodeName)
+		_, err := createExternalNodeCRD(data, vm.nodeName, vm.ifName, vm.ip)
+		require.Nil(t, err, "Failed to create ExternalNode")
+		startAntreaAgent(t, data, vm)
+	}
+	return vmList, nil
+}
+
+// teardownVMAgentTest deletes ExternalNode, verifies ExternalEntity is deleted
+// and verifies uplink configuration is restored.
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
+			var tempVm vmInfo
+			if vm.osType == "Linux" {
+				tempVm = getVMInfo(t, data, vm.nodeName)
+			} else {
+				tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if vm.ifName != tempVm.ifName {
+				t.Errorf("Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+				return false, nil
+			}
+			if vm.ip != tempVm.ip {
+				t.Errorf("Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			assert.Fail(t, "Failed to verify uplink configuration after cleanup")
+			return
+		}
+	}
+	t.Logf("TestVMAgent teardown")
+	for _, vm := range vmList {
+		hash := sha1.New() // #nosec G401: not used for security purposes
+		io.WriteString(hash, vm.ifName)
+		hashedIfName := hex.EncodeToString(hash.Sum(nil))
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
+		err := data.crdClient.CrdV1alpha1().ExternalNodes(nameSpace).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{})
+		require.Nil(t, err, "Failed to delete ExternalNode")
```

Comment:
that probably shouldn't be a `require` assertion, we want to keep on cleaning even if one deletion fails

Reply:
Done. I will just dump a log message here.